### PR TITLE
[AIRFLOW-3346] Add hook and operator for GCP transfer service.

### PR DESCRIPTION
**Pull request by jmcarp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3346
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Add a hook and operator(s) to connect to gcp storage transfer service and transfer files from s3 to gcp (and gcp to gcp) without copying to local disk.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/hooks/gcp_transfer_hook.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from googleapiclient.discovery import build
+
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+# Time to sleep between active checks of the operation results
+TIME_TO_SLEEP_IN_SECONDS = 1
+
+
+# noinspection PyAbstractClass
+class TransferServiceHook(GoogleCloudBaseHook):
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_transfer_hook.py]: Rename `TransferServiceHook` to `GcpTransferServiceHook`

File: airflow/contrib/operators/s3_to_gcs_transfer_operator.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    Synchronizes an S3 bucket with a Google Cloud Storage bucket using the
+    GCP Storage Transfer Service.
+
+    :param s3_bucket: The S3 bucket where to find the objects. (templated)
+    :type s3_bucket: str
+    :param gcs_bucket: The destination Google Cloud Storage bucket
+        where you want to store the files. (templated)
+    :type gcs_bucket: str
+    :param project_id: The GCP project ID
```
Comment (kaxil) [inline airflow/contrib/operators/s3_to_gcs_transfer_operator.py]: Reword it to `The ID of the Google Cloud Platform Console project that owns the job.`

File: tests/contrib/hooks/test_gcp_transfer_hook.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+try:
+    from unittest import mock
+except ImportError:
+    try:
+        import mock
+    except ImportError:
+        mock = None
+
+
+class TestTransferServiceHook(unittest.TestCase):
```
Comment (kaxil) [inline tests/contrib/hooks/test_gcp_transfer_hook.py]: `TestTransferServiceHook` -> `TestGcpTransferServiceHook `

**Review by kaxil — CHANGES_REQUESTED** [review]
Minor suggestion.

Also, you need to add this to `integrations.rst` and `code.rst`.

@potiuk Would be good to get your opinion as well.

**Comment by jmcarp** [comment]
Thanks for the update @kaxil! Let me know when you're done and I'll squash commits.

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
I have squashed and merged it. Thanks @jmcarp for your contribution :)
